⚠️(0.7.1) Removal of macros from TimingData.h (REPLACED BY #499, WONT REOPEN)

sukibaby opened the pull request:
Un-does a set of complicated macros. The reasons are that this allows the compiler to better optimize than it would as a macro, and many errors were found to be hiding behind that macro.

Because this replaces the macro with a large section of visually similar code handling different forms of timing segments, a comment has been added explaining the purpose of the functions.

sukibaby committed:
Remove DefineSegment macros from TimingData.h

Un-does a complicated macro. The reasons are that this allows the compiler to better optimize than it would as a macro, and many errors were found to be hiding behind that macro.

Because this replaces the macro with a large section of visually similar code handling different timing segments within the TimingData class, a comment has been added explaining the purpose of the functions.

sukibaby commented:
Re-pushed with fixed spacing to match what was done in the TimingSegmentToX macros PR

jsirex commented on src/TimingData.h:
```diff
@@ -193,50 +193,313 @@ class TimingData
+	
+	void AddSegment(const DelaySegment& seg)
+	{
+		AddSegment(&seg);
```
Can't understand this code. Is it a endless recursion? Was macro expected to be inside of  NotesLoader class?

teejusb replied:
It's not endless recursion because it's calling the pointer variant `AddSegment(const TimingSegment* seg)` although I admit it is a bit confusing. Might be worth a name change.

sukibaby replied:
This is not too uncommon in the code base unfortunately. There are a lot of variants of functions that share the same name but differ by what arguments they take.  It could definitely be improved for readability.  

With that said, this PR is just the expanded macro and nothing else. If @teejusb wants them renamed in this PR I'll do that though.

sukibaby commented:
I've added a comment block explaining these functions to make it less obtuse. 

I'm a little worried about renaming the `AddSegment` functions if we can help it, because it works as-is and it would be a lot of effort to correlate all references to its variant and test.

sukibaby committed:
Remove TimingSegment macros from TimingData.h

Un-does a complicated macro. The reasons are that this allows the compiler to better optimize than it would as a macro, and many errors were found to be hiding behind that macro.

Because this replaces the macro with a large section of visually similar code handling different timing segments pointers, a comment has been added explaining the purpose of the functions.

sukibaby commented:
Combined #325  in here.

teejusb commented on src/TimingData.h:
```diff
@@ -27,36 +27,182 @@ inline T ToDerived( const TimingSegment *t, TimingSegmentType tst )
 }
 
-#define TimingSegmentToXWithName(Seg, SegName, SegType) \
-	inline const Seg* To##SegName( const TimingSegment *t ) \
```
I think you can create inline templated functions here, which I think is nicer.

sukibaby replied:
> Yeah, I'd also like that a lot better. Based on this and your comment on the other macro changes PR, and the age of this PR (it would cause conflicts in notefield if i update it now) I will make a new one that's better organized and has templates instead of expanded inlines.

_Originally posted by @sukibaby in https://github.com/itgmania/itgmania/pull/400#discussion_r1769467262_

sukibaby commented:
I wanted to make a new PR to leave the 0.7.1 changes alone in case we need to refer to this later for whatever reason. Please see the new PR #499.
